### PR TITLE
Fix Brotli compression

### LINKED ISSUE #125: webkitRequestFileSystem and brotli_wasm_bg.wasm problem
Thanks for reporting an issue!

### Steps to reproduce
Add package to NextJS 14 project

### Expected behaviour
Package imported correctly

### Actual behaviour
Package is throwing error on NextJS build

### Any logs, error output, etc?
Error is either:
```
Error: ENOENT: no such file or directory, open '/Users/<projectPath>/.next/server/chunks/brotli_wasm_bg.wasm'
```
or
```
 Collecting page data  ...TypeError: Cannot read properties of undefined (reading 'webkitRequestFileSystem')
```
depending on which import statement I use:
```
import parquetjs from "@dsnp/parquetjs";
import parquetjs from "@dsnp/parquetjs/dist/browser/parquet.esm";
```

### Any other comments?
...

### LINKED ISSUE #140: Error occurs when parsing Parquet files using Brotli compression
### Steps to reproduce

Use the below code to parse a Parquet file using Brotli compression.

### Expected behaviour

The parse is successful.

### Actual behaviour

An error occurs:

> la.__wbindgen_malloc is not a function

Note that this does not happen for other supported compression methods.

### Any logs, error output, etc?

This is the error:

```json
{
  "name": "TypeError",
  "message": "la.__wbindgen_malloc is not a function",
  "stack": "TypeError: la.__wbindgen_malloc is not a function\n    at passArray8ToWasm (/Applications/Microsoft Azure Storage Explorer.app/Contents/Resources/app/node_modules/@storage-explorer/file-preview/dist/src/index.js:27:2264713)\n    at ju.exports.decompress (/Applications/Microsoft Azure Storage Explorer.app/Contents/Resources/app/node_modules/@storage-explorer/file-preview/dist/src/index.js:27:2265227)\n    at Object.inflate_brotli [as inflate] (/Applications/Microsoft Azure Storage Explorer.app/Contents/Resources/app/node_modules/@storage-explorer/file-preview/dist/src/index.js:27:2267375)\n    at Object.inflate (/Applications/Microsoft Azure Storage Explorer.app/Contents/Resources/app/node_modules/@storage-explorer/file-preview/dist/src/index.js:27:2267061)\n    at decodeDictionaryPage (/Applications/Microsoft Azure Storage Explorer.app/Contents/Resources/app/node_modules/@storage-explorer/file-preview/dist/src/index.js:72:280198)\n    at decodePage (/Applications/Microsoft Azure Storage Explorer.app/Contents/Resources/app/node_modules/@storage-explorer/file-preview/dist/src/index.js:72:279268)\n    at /Applications/Microsoft Azure Storage Explorer.app/Contents/Resources/app/node_modules/@storage-explorer/file-preview/dist/src/index.js:72:277570\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async e.readColumnChunk (/Applications/Microsoft Azure Storage Explorer.app/Contents/Resources/app/node_modules/@storage-explorer/file-preview/dist/src/index.js:72:277516)\n    at async e.readRowGroup (/Applications/Microsoft Azure Storage Explorer.app/Contents/Resources/app/node_modules/@storage-explorer/file-preview/dist/src/index.js:72:276909)"
}
```

### Any other comments?

This occurs within the context of another app. I don't know for sure if this happens "out of the box" or not. Could this be due to a bad or missing Rust binding in `brotli-wasm`? Here's the Typescript code that is used:

```typescript
import { ParquetReader } from "@dsnp/parquetjs";

async function* generateRecord(reader: ParquetReader): AsyncGenerator<any[]> {
  const cursor = reader.getCursor();
  let record: any;
  while (record = await cursor.next()) {
    yield record;
  }
}

const buffer = ...; // Retrieve by downloading from some URL or reading from local file.

const reader = await ParquetReader.openBuffer(buffer));
const schema = reader.getSchema();
const columns = Object.keys(schema.fields);
const rows = [];

for await (const record of generateRecord(this.#reader)) {
  const rowArray = columns.reduce<TabularCellValue[]>((acc, key) => {
    const value = record[key];
    if (value === null || value === undefined) {
      acc.push(undefined);
    } else if (value instanceof Date) {
      acc.push({ type: "Date", value: value.getTime() });
    } else if (typeof value === "bigint") {
      acc.push({ type: "BigInt", value: value.toString() });
    } else if (typeof value === "object") {
      acc.push(JSON.stringify(value, (_, v) => typeof v === "bigint" ? v.toString() : v));
    } else if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
      acc.push(value);
    } else {
      acc.push(undefined);
    }

    return acc;
  }, []);
  rows.push(rowArray);
}

...
```

### PR DESCRIPTION
**Pull request by wilwade**
Note: The reference Brotli file doesn't work because it is too large. It decompresses to an extremely large size (1Gb+)

Closes #140
Closes #125 (Likely?)

A new brotli sample file was generated using python code seen below.

## What Changed?

- Moved the esbuild work to mjs. This was required to get some of the builds working correctly
- Fixed bug in the esbuild when it tried to build the browser code in parallel with the test code causing a race condition
- Split the compression.ts file into a browser version and a node version
- Swapped over to use `esbuild-plugin-wat` which worked better than the copy-pasted one from esbuild
- Integrated the brotli-wasm correctly for browser, but used brotli natively in nodejs

## Testing!

- There is a nodejs test for the node version
- There is a browser test for it as well:
    1. `npm i`
    2. `npm run build:browser`
    3. `npx serve .`
    4. `open http://localhost:3000/test/browser/` in your preferred browser
- The example server also has it:
    - cd examples/server/
    - npm i
    - node app.js
    - cd ../../ && npm run serve
    - `open http://localhost:3000/` in your preferred browser

### Brotli Sample File generation script

```python
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq

# Create a small sample DataFrame
data = {
    'id': [1, 2, 3, 4, 5],
    'name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve'],
    'age': [25, 30, 35, 40, 45]
}

df = pd.DataFrame(data)

# Convert DataFrame to PyArrow Table
table = pa.Table.from_pandas(df)

# Define output Parquet file path
output_file = "sample_brotli_compressed.parquet"

# Write to Parquet file with Brotli compression
pq.write_table(table, output_file, compression='BROTLI')

print(f"File {output_file} created successfully!")
``

### CODE REVIEW

File: esbuild-serve.mjs
Comment (wilwade) [inline esbuild-serve.mjs:1]: `esbuild-plugin-wat` does what we did, but better!

File: examples/server/views/parquetFiles.ejs
Comment (wilwade) [inline examples/server/views/parquetFiles.ejs:1]: Visible output!

File: lib/browser/brotli.ts
Comment (wilwade) [inline lib/browser/brotli.ts:1]: This code is mostly copied from https://github.com/httptoolkit/brotli-wasm?tab=readme-ov-file#in-browser-with-streams

File: esbuild.mjs [hunk 1]
```diff
@@ -10,11 +10,11 @@ const baseConfig = {
     'process.env.NODE_ENV': '"production"',
     global: 'window',
   },
-  inject: ['./esbuild-shims.js'],
-  minify: true,
+  inject: ['./esbuild-shims.mjs'],
+  minify: false,
```
Comment (shannonwells) [inline esbuild.mjs:14]: don't we want to minify?

Comment (wilwade) [reply esbuild.mjs:14]: Good catch. I had disabled it for testing.
```suggestion
  minify: true,
```

File: esbuild-serve.mjs [hunk 2]
```diff
@@ -3,24 +3,27 @@
...
   .context({
     entryPoints: ['parquet.ts'],
     outfile: 'main.js',
     define: { 'process.env.NODE_DEBUG': 'false', 'process.env.NODE_ENV': '"production"', global: 'window' },
     platform: 'browser',
-    plugins: [compressionBrowserPlugin, wasmPlugin],
+    plugins: [compressionBrowserPlugin, watPlugin()],
     sourcemap: 'external',
     bundle: true,
+    minify: false,
```
Comment (wilwade) [inline esbuild-serve.mjs:19]: Keeping minify off here for the local serve.

**Review by shannonwells — APPROVED** [review]
I ran the examples and reviewed the code.  Looks good to go, glad you found it was mainly a poor example file.

**Pull request merged**